### PR TITLE
Introduce type precedence and type inference

### PR DESCRIPTION
**Pull request by mxm**
Type precedence during casting enables us to decide if parameters in expressions
can be safely cast to another parameter's type. This is possible by assigning
IDs to all the types which denote the rank of the types. Higher IDs denote
higher precedence.

All the casts in the ExpressionAnalyzer now follow the same
pattern: When looking up a function, we use the FunctionResolver interface which
may return new input types through the getSignature(List<DataType>) method. The
analyzer can then cast the current arguments to match the types to the
FunctionImplementation.

The type matching works through the FuncParams class. Each FuncParams is
composed of zero, one, or multiple Params which contain the valid types for the
parameters. A Param can be passed to FuncParams multiple types to connect
parameters with each other. For example, FuncParams.of(a, a, b) means that a is
of the same type. This information allows us to improve type inference.

The type inference works as follows. First, a Param is unbound. When arguments
are passed, it becomes bound. When it is bound, it will try to convert an input
to its current type, the input type, or one of the valid types provided for the
Param. For example, if we first bind to Integer and we get a Long, we convert to
Long; if we then get a String and String is a valid type, we will convert to
String.

Due to these type handling improvements, we can now avoid column casting. We
cast Literal values but require explicit casts whenever two columns are compared
which don't have a matching type. Column casts should generally be avoided
because they can slow down Lucene queries. Type precedence always has preference
when converting types but may be omitted if conversion can still be
performed (i.e. column has lower precedence).

Supersedes #6301.

### CODE REVIEW

File: core/src/main/java/io/crate/types/CollectionType.java [hunk 1]
```diff
@@ -72,10 +72,16 @@ public void writeTo(StreamOutput out) throws IOException {
         DataTypes.toStream(innerType, out);
     }
 
+    @Override
     public String getName() {
-        return innerType.getName();
+        String innerName = "";
+        if (innerType != null) {
```
Comment (mfussenegger) [inline core/src/main/java/io/crate/types/CollectionType.java]: Can we enforce the innerType to be nonnull? `innerType()` isn't annotated as nullable, `isNumeric`, `isDecimal`, `compareTo` etc.  lack null checks.

File: core/src/main/java/io/crate/types/DataType.java [hunk 2]
```diff
@@ -43,6 +79,20 @@
 
     public abstract int compareValueTo(T val1, T val2);
 
+    @Nullable
+    public DataType innerType() {
```
Comment (mfussenegger) [inline core/src/main/java/io/crate/types/DataType.java]: Why do we need a `innerType` on the base DataType class?

File: core/src/main/java/io/crate/types/SingleColumnTableType.java [hunk 3]
```diff
@@ -57,7 +62,11 @@ public String getName() {
         if (value == null) {
             return null;
         }
-        return (Object[]) value;
+        Object[] array = (Object[]) value;
+        for (int i = 0; i < array.length; i++) {
+            array[i] = innerType.value(array[i]);
```
Comment (mfussenegger) [inline core/src/main/java/io/crate/types/SingleColumnTableType.java]: Why is this not necessary on master?

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 4]
```diff
@@ -202,40 +200,15 @@ public Symbol generateQuerySymbol(Optional<Expression> whereExpression, Expressi
...
         for (Expression expression : node.getArguments()) {
             Symbol argSymbol = expression.accept(innerAnalyzer, context);
-
-            argumentTypes.add(argSymbol.valueType());
             arguments.add(argSymbol);
         }
 
         List<String> parts = node.getName().getParts();
-        String schema = null;
+        String schema = transactionContext.sessionContext().defaultSchema();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: Does this still have the same semantics if we always set the schema from the session? Afaik we wanted to treat `x()` with session schema `foo` differently than `foo.x()` (due to built-ins not being in any schema)

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionToByteSizeValueVisitor.java [hunk 5]
```diff
@@ -62,7 +62,7 @@ protected ByteSizeValue visitLongLiteral(LongLiteral node, Row context) {
 
     @Override
     protected ByteSizeValue visitDoubleLiteral(DoubleLiteral node, Row context) {
-        return new ByteSizeValue(((Double) node.getValue()).longValue());
+        return new ByteSizeValue(Double.valueOf(node.getValue()).longValue());
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/expressions/ExpressionToByteSizeValueVisitor.java]: Minor, but would be nice to avoid unrelated style changes like these

File: sql/src/main/java/io/crate/analyze/symbol/Symbol.java [hunk 6]
```diff
@@ -42,4 +43,12 @@ public static boolean isLiteral(Symbol symbol, DataType expectedType) {
     public String toString() {
         return representation();
     }
+
+    /**
+     * Typically, we only allow casting of Literals and undefined types.
+     */
+    @Override
+    public boolean canBeCasted() {
+        return valueType() != null && valueType().id() == UndefinedType.INSTANCE.id();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/analyze/symbol/Symbol.java]: imho valueType shouldn't be nullable. So far we've never any null checks if we use valueType

File: sql/src/main/java/io/crate/metadata/FunctionResolver.java [hunk 7]
```diff
@@ -1,57 +1,60 @@
...
     @Nullable
-    List<DataType> getSignature(List<DataType> dataTypes);
+    List<DataType> getSignature(List<? extends FuncArg> funcArgs);
+
+    /**
+     * Returns the actual function implementation for the given argument type list.
+     *
+     * @throws java.lang.IllegalArgumentException thrown if there is no function that can handle the given types.
+     */
+    FunctionImplementation getForTypes(List<DataType> symbols) throws IllegalArgumentException;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/FunctionResolver.java]: Any reason this was moved?

File: sql/src/main/java/io/crate/metadata/Functions.java [hunk 8]
```diff
@@ -143,24 +197,38 @@ public FunctionImplementation getQualified(FunctionIdent ident) throws Unsupport
         return impl;
     }
 
-    public static UnsupportedOperationException createUnknownFunctionException(String name, List<DataType> argumentTypes) {
+    public static UnsupportedOperationException createUnknownFunctionExceptionFromArgs(String name,
+                                                                                       List<? extends FuncArg> arguments) {
+        List<DataType> dataTypes = arguments.stream().map(FuncArg::valueType).collect(Collectors.toList());
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/Functions.java]: Could use `Lists2.copyAndReplace` as alternative with less allocation overhead.

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            try {
+                return this.funcParams;
+            } finally {
+                // prevent reuse of the Builder
+                this.funcParams = null;
+            }
+        }
+    }
+
+    public static Builder of(Param... parameters) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: I'd rename these methdos to `builder` to be consistent with other places in the code (`Settings.builder()`, `ClusterState.builder()`, etc.. )

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+            this.funcParams.varArgsIndependent = true;
+            return this;
+        }
+
+        public FuncParams build() {
+            try {
+                return this.funcParams;
+            } finally {
+                // prevent reuse of the Builder
+                this.funcParams = null;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: Just curious: Is this pattern more efficient than using a temporary local variable?

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+        this.varArgParameters = new Param[]{};
+        this.maxVarArgOccurrences = -1;
+    }
+
+    public static class Builder {
+
+        private FuncParams funcParams;
+
+        public Builder(Param... parameters) {
+            this.funcParams = new FuncParams(parameters);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: It looks like the builder is mostly used during node-startup and not at runtime - wouldn't it be cleaner to make all FuncParams attributes final and have the Builder have it's own mutable attributes?

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+    private void resetBoundTypes() {
+        for (Param parameter : parameters) {
+            parameter.clear();
+        }
+        for (Param parameter : varArgParameters) {
+            parameter.clear();
+        }
+    }
+
+    private void bindParams(List<? extends FuncArg> params) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: I guess this method could use some explanation

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+ *
+ *      FuncParams.of().withVarArgs(2, Param.of(INTEGER)).build()
+ *      // => Maximum of two variable integer arguments.
+ *
+ *      FuncParams.of().withIndependentVarArgs(Param.of(ANY)).build()
+ *      // => Unlimited number of arguments which don't have to be of the same type
+ * }
+ *
+ *
+ * This class is thread-safe and can be accessed by multiple threads at once.
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: the `parameters` are being mutated, so this is only thread-safe if the `Param` class provides type-safety. I think that should be mentioned (and Param should probably become a final class, to prevent subclassing)

And I'm not sure if this really is thread-safe. E.g if `resetBoundTypes` is called concurrently from multiple threads, this wouldn't work?

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+                DataType innerType = this.innerType.getBoundType(multiBind);
+                if (innerType != null) {
+                    return ((CollectionType) boundType).newInstance(innerType);
+                }
+            }
+        }
+        return boundType;
+    }
+
+    void bind(FuncArg funcArg, boolean multiBind) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/functions/params/Param.java]: This could also use some explanation - or maybe split it up a bit?

File: sql/src/main/java/io/crate/operation/operator/OperatorModule.java [hunk 15]
```diff
@@ -57,11 +55,13 @@ protected void configure() {
         // bind all registered functions and resolver
         // by doing it here instead of the register functions, plugins can also use the
         // register functions in their onModule(...) hooks
-        functionBinder = MapBinder.newMapBinder(binder(), FunctionIdent.class, FunctionImplementation.class);
-        dynamicFunctionBinder = MapBinder.newMapBinder(binder(), String.class, FunctionResolver.class);
+        MapBinder<FunctionIdent, FunctionImplementation> functionBinder =
+            MapBinder.newMapBinder(binder(), FunctionIdent.class, FunctionImplementation.class);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/operator/OperatorModule.java]: This change is unrelated?

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java (hunk 11 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: Or just provide the appropriate constructors in `FuncParams` - it look like the methods are never really chained?

File: sql/src/main/java/io/crate/operation/scalar/regex/RegexMatcher.java [hunk 16]
```diff
@@ -126,8 +126,12 @@ public static int parseFlags(@Nullable String flagsString) {
                 case 'd':
                     flags = flags | Pattern.UNIX_LINES;
                     break;
-                default:
+                case ' ':
+                case 'g':
+                    // handled in isGlobalFunction
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/scalar/regex/RegexMatcher.java]: unrelated change?

File: sql/src/test/java/io/crate/operation/operator/OrOperatorTest.java [hunk 17]
```diff
@@ -11,13 +11,12 @@
 
     @Test
     public void testNormalize() throws Exception {
-        assertNormalize("name or true", isLiteral(true));
-        assertNormalize("true or name", isLiteral(true));
-        assertNormalize("false or name", isField("name"));
-        assertNormalize("name or name", isFunction(OrOperator.NAME));
-
-        assertNormalize("true or 1/0", isLiteral(true));
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/operator/OrOperatorTest.java]: will this now cause an error? This would be another breaking change and I think one that we should not introduce.

File: core/src/main/java/io/crate/types/DataType.java (hunk 2 above)
Comment (mxm) [reply core/src/main/java/io/crate/types/DataType.java]: To have a consistent way to solve the inner type of a DataType. Otherwise we need `instanceof`.

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java (hunk 12 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: My opinion is that without reading the method body, just by reading the signature/docs I should be able to guess what a method is doing - in this case (and `bind`) that's not the case.

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java (hunk 10 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]:
```
result = this.funcParams;
this.funcParams = null;
return result;
```

It's the same as using finally - just asking if there is any advantage to the finally pattern.

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java (hunk 11 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: > The Builder doesn't cause runtime overhead. This is a really minor issue. I'd rather keep the API readable for other people then optimizing one object allocation upon startup.

That's kinda my point - I'd be in favor of *not* optimizing the object allocation and properly decouple the Builder attributes from the FuncParams and then create the FuncParams with all fields final.

If you're reading code and try to figure out if something really *is* thread-safe, it's much easier if the attributes are marked as final vs. "effectively" final.

File: sql/src/test/java/io/crate/operation/operator/OrOperatorTest.java (hunk 17 above)
Comment (mfussenegger) [reply sql/src/test/java/io/crate/operation/operator/OrOperatorTest.java]: You're right, but `true or (1/0 = 10)` should still return `true` and not result in a failure

File: core/src/main/java/io/crate/types/DataType.java (hunk 2 above)
Comment (mfussenegger) [reply core/src/main/java/io/crate/types/DataType.java]: I'd be in favor of the `instanceof` instead of adding a property that doesn't make sense for the majority of the types. E.g. if we add a `RowType` or extend the ObjectType to know it's inner types, this will be even more confusing because it would have many inner types, not only one.

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java (hunk 4 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: Can we extract this? If it's a bugfix we also need to backport it

File: sql/src/main/java/io/crate/operation/scalar/regex/RegexMatcher.java (hunk 16 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/operation/scalar/regex/RegexMatcher.java]: Please do extract it - this probably also warrants a changes entry

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java (hunk 14 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: I think `bindParam` and `rebindParam` would make this more clear (and might even help the JIT to inline - though that probably doesn't matter here)

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java (hunk 10 above)
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: Ah, hadn't thought about that. I guess not really any difference.

**Comment by mxm** [comment]
Ah, get it now. I agree it is easier to reason about at a glance. Will
adapt.

**Comment by mxm** [comment]
@mfussenegger I've compiled the results of our discussion in the fixup commits.

File: sql/src/main/java/io/crate/operation/operator/CmpOperator.java [hunk 18]
```diff
@@ -55,4 +63,35 @@ public Boolean evaluate(Input<Object>... args) {
...
+        CmpResolver(String name) {
+            this(name, FuncParams.builder(primitiveTypes, primitiveTypes).build());
+        }
+
+        CmpResolver(String name, FuncParams funcParams) {
+            super(funcParams);
+            this.name = name;
+        }
+
+        protected abstract FunctionImplementation createInstance(FunctionInfo info);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/operator/CmpOperator.java]: Might want to consider changing the constructor to take a `Function<FunctionInfo, FunctionImplementation>` and make this non-abstract. Would allow to use method refs (Ex. `new CmpResolver(NAME, GteOperator::new)` (and satisfy my personal composition > inheritance preference :smile: )

File: sql/src/test/java/io/crate/operation/scalar/TypeInferenceTest.java [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+package io.crate.operation.scalar;
+
+import org.junit.Test;
+
+public class TypeInferenceTest extends AbstractScalarFunctionsTest {
+
+    @Test
+    public void testCoalesce() throws Exception {
+        assertEvaluate("coalesce(null, 1::integer, 2::long, 3.0)", 1.0);
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/scalar/TypeInferenceTest.java]: Could we also add a test with a column that shows that integer casts a no longer necessary for numeric literals?

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+        Objects.requireNonNull(funcArg, "funcArg to bind must not be null");
+        final FuncArg bound = this.boundType.get();
+        final FuncArg updatedType;
+        if (bound == null) {
+            updatedType = bindFresh(funcArg, multiBind);
+        } else {
+            updatedType = rebind(bound, funcArg);
```
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: `unbind` is **always** called in `FuncParams` before binding new arguments to parameters. `rebind` is called when a `Param` is already bound and needs to be `rebound`.

For example, func(a, b, a) with args (1::int, 'foo'::string, 2::long):

- unbind() => (unbound, unbound, unbound)
- bind(1::int) => (int, unbound, int)
- bind('foo'::string) => (int, string, int)
- bind(2::long) => (long, string, long)

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+            for (DataType targetType : validTypes) {
+                convertedType = convert(funcArg, targetType);
+                if (convertedType != null) {
+                    break;
+                }
+            }
+            if (convertedType == null) {
+                throw new ConversionException(funcArg, validTypes);
+            }
+            return convertedType;
```
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: Sure it is because converting is not always possible.

[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: Please explain why you think this is not reachable.

File: sql/src/main/java/io/crate/operation/operator/CmpOperator.java (hunk 18 above)
Comment (mxm) [reply sql/src/main/java/io/crate/operation/operator/CmpOperator.java]: Alright, whatever makes you happy 😎

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: Ah nevermind, I read the `break` as return.

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: Would be cool to integrate such an example in the code as comment

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+    private final Param innerType;
+
+    private Param(DataType... validTypes) {
+        this(Collections.emptyList(), validTypes, null);
+    }
+
+    private Param(Collection<DataType> validTypes, DataType[] validTypes2, @Nullable Param innerType) {
+        this.boundType = new ThreadLocal<>();
+        this.validTypes = new TreeSet<>((o1, o2) -> {
+            if (o1.precedes(o2)) {
```
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: 👍

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+    }
+
+    private void checkForRemainingParams(List<? extends FuncArg> params, int pos) {
+        int numberOfParameters = params.size();
+        if (pos != numberOfParameters) {
+            final int numArgsToPrint;
+            if (varArgParameters.length == 0) {
+                numArgsToPrint = 1;
+            } else if (maxVarArgOccurrences == UNLIMITED_VAR_ARGS) {
+                numArgsToPrint = 5;
```
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: 5 seems like a sane number for the number of example inputs.

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+     * Create a FuncParam with a fixed Signature.
+     * @param fixedSignature The {@link DataType}s to accept.
+     * @return A FuncParam accepting the supplied signature.
+     */
+    public static FuncParams of(List<DataType> fixedSignature) {
+        Param[] params = new Param[fixedSignature.size()];
+        for (int i = 0; i < params.length; i++) {
+            params[i] = Param.of(fixedSignature.get(i));
+        }
+        return new FuncParams(params, new Param[]{}, UNLIMITED_VAR_ARGS, false);
```
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: It was your suggestion to use only final fields and have everything in the constructor. This is the default initialization. It is a private constructor. If you don't set any var args `Param`s, it doesn't matter how often they occur. If you pass any var args, we will raise an error.

File: sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java [hunk 25]
```diff
@@ -880,40 +866,85 @@ protected Symbol visitSubqueryExpression(SubqueryExpression node, ExpressionAnal
...
+        Function newFunction = new Function(functionInfo, castArguments);
         if (functionInfo.isDeterministic() && Symbols.allLiterals(newFunction)) {
-            FunctionImplementation funcImpl = functions.getQualified(newFunction.info().ident());
             return funcImpl.normalizeSymbol(newFunction, transactionContext);
         }
         return newFunction;
     }
 
+
+    private static FunctionImplementation getUdfFunction(@Nullable String schemaProvided,
```
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java]: 👍

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: done

File: sql/src/test/java/io/crate/operation/scalar/TypeInferenceTest.java (hunk 19 above)
Comment (mxm) [reply sql/src/test/java/io/crate/operation/scalar/TypeInferenceTest.java]: I've added a test in `FuncParamsTest`. Adding another one to test this end-to-end.

File: core/src/main/java/io/crate/types/CollectionType.java [hunk 26]
```diff
@@ -119,13 +125,15 @@ public boolean equals(Object o) {
         if (!super.equals(o)) return false;
 
         CollectionType collectionType = (CollectionType) o;
-        return innerType.equals(collectionType.innerType);
+        return innerType != null && innerType.equals(collectionType.innerType);
```
Comment (mfussenegger) [inline core/src/main/java/io/crate/types/CollectionType.java]: Isn't `innerType` now non-nullable again?

File: core/src/main/java/io/crate/types/CollectionType.java [hunk 27]
```diff
@@ -119,13 +125,15 @@ public boolean equals(Object o) {
...
     }
 
     @Override
     public int hashCode() {
         int result = super.hashCode();
-        result = 31 * result + innerType.hashCode();
+        if (innerType != null) {
```
Comment (mfussenegger) [inline core/src/main/java/io/crate/types/CollectionType.java]: Same here

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: Maybe leave a short comment along the lines of "5 is an arbitrary limit to keep error msg length reasonable"

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: I saw that you changed this but reverted this, what was the reason? (It's okay, just curious)

File: core/src/main/java/io/crate/types/CollectionType.java (hunk 26 above)
Comment (mxm) [reply core/src/main/java/io/crate/types/CollectionType.java]: It shouldn't be null, added a Precondition in the constructor and removing the check.

File: sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java]: Extracted this to a variable to make it more clear.

File: sql/src/main/java/io/crate/metadata/functions/params/Param.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: Forgot to drop you a comment. I removed this because I need the valid types in descending precedence order. When I use the default compare to, I would have to negate the result. Maybe that's still easier to read but I didn't want to fiddle with it too much.

[the remark this answers is not in the document]
Comment (mxm) [reply sql/src/main/java/io/crate/metadata/functions/params/Param.java]: This would work:
```java
new Comparator<DataType>() {
        @Override
        public int compare(DataType o1, DataType o2) {
            return o1.precedence().compareTo(o2.precedence());
        }
}.reversed());
```
